### PR TITLE
added Left and Right Circular Projection matrix

### PR DESCRIPTION
**Pull request by JAMSADIQ**
Added a new function in this code for left and right circular polarization projection matrix.

### CODE REVIEW

**Comment by spxiwh** [comment]
@a-r-williamson I've assigned this one to you to look over.

**Comment by JAMSADIQ** [comment]
With the above commits now pycbc-multi-inspiral and lalapps_coh_PTF_inspiral  code give same results when using left anf right circular polarization matrices. The details can be found at https://wiki.ligo.org/Bursts/CircularlyPolarizedProjectionMatrix
In order to check the code for accepting merge request my modified code can be used for an injection test with command line options given in https://wiki.ligo.org/Bursts/CircularlyPolarizedProjectionMatrix page.

File: bin/pycbc_multi_inspiral [hunk 1]
```diff
@@ -107,6 +107,7 @@ parser.add_argument("--null-step", type=float, default=20., help="""
 parser.add_argument("--trigger-time", type=int, help="""
                     Time of the GRB, used to set the antenna patterns.""")
 
+parser.add_argument("--projection-type", default='standard', type=str, help="""Choice of projection matrix with standard, left circular or right circular""")
```
Comment (a-r-williamson) [inline bin/pycbc_multi_inspiral:110]:
```suggestion
parser.add_argument("--projection-type", default="standard",
                    choices=["standard", "left", "right", "left+right"],
                    help="Choice of projection matrix. 'Left' and 'right' "
                         "correspond to face-away and face-on.")
```

File: bin/pycbc_multi_inspiral [hunk 2]
```diff
@@ -211,6 +211,16 @@ def get_projection_matrix(wp, wc):
     return projection_matrix
 
 
+def circular_projection_matrix(wp, wc):
```
Comment (a-r-williamson) [inline bin/pycbc_multi_inspiral:214]: Can we perhaps unify this into the above function `get_projection_matrix` which could have a kwarg for the projection name?

File: bin/pycbc_multi_inspiral [hunk 3]
```diff
@@ -513,6 +528,10 @@ with ctx:
             sigma = {ifo : np.sqrt(sigmasq[ifo]) for ifo in opt.instruments}
             # Every time s_num is zero or we skip the segment, we run new 
             # template to increment the template index
+            ####### Testing values
```
Comment (a-r-williamson) [inline bin/pycbc_multi_inspiral:531]:
```suggestion
```

File: bin/pycbc_multi_inspiral [hunk 4]
```diff
@@ -513,6 +528,10 @@ with ctx:
...
+            wp,wc=get_weighted_antenna_patterns(Fp,Fc,sigma)
+            projection_matrix = get_projection_matrix(wp,wc)
+            LCprojection_matrix, RCprojection_matrix = circular_projection_matrix(wp, wc)
```
Comment (a-r-williamson) [inline bin/pycbc_multi_inspiral:534]: We also need to account for the normal GRB search behaviour where both right and left projections are used.

**Comment by JAMSADIQ** [comment]
I make changes and push them. Kindly check it

File: bin/pycbc_multi_inspiral [hunk 5]
```diff
@@ -197,18 +195,61 @@ def get_weighted_antenna_patterns(Fp_dict, Fc_dict, sigma_dict):
...
+                            % str(projection_type))
+        return right_projection_matrix
+
+    elif projection_type=='left+right':
+        logging.info("choice of projection matrix is %s"
+                            % str(projection_type))
+        return left_projection_matrix, right_projection_matrix
+
+    else:
+        logging.info("choice of projection matrix is not available")
```
Comment (titodalcanton) [inline bin/pycbc_multi_inspiral:240]: This should probably be an exception, not a `logging.info()` call:
```python
raise ValueError('unknown projection type ' + str(projection_type))
```

File: bin/pycbc_multi_inspiral [hunk 6]
```diff
@@ -197,18 +195,61 @@ def get_weighted_antenna_patterns(Fp_dict, Fc_dict, sigma_dict):
...
+    left_projection_matrix = (np.outer(wp, wp) + np.outer(wc, wc) +
+                              (np.outer(wp, wc) - np.outer(wc, wp))
+                              *1j) / side_denominator
+    right_projection_matrix = (np.outer(wp, wp) + np.outer(wc, wc) +
+                               (np.outer(wc, wp) - np.outer(wp, wc))
+                               *1j) / side_denominator
+ 
+    if projection_type=='standard':
+        logging.info("choice of projection matrix is %s"
+                            % str(projection_type))
```
Comment (titodalcanton) [inline bin/pycbc_multi_inspiral:221]: Calls to `logging.info()` should not use the string formatting operations (`%` and `.format()`). Instead, do this:
```python
logging.info("choice of projection matrix is %s", projection_type)
```
(this comment also applies to `logging.info()` further below).

**Comment by JAMSADIQ** [comment]
Fixed some issues for review.

File: bin/pycbc_multi_inspiral [hunk 7]
```diff
@@ -197,20 +200,33 @@ def get_weighted_antenna_patterns(Fp_dict, Fc_dict, sigma_dict):
     return wp, wc
 
 
-def get_projection_matrix(wp, wc):
+def get_projection_matrix(wp, wc, projection='standard'):
     """
     Output: projection_matrix: Projects the data onto the signal space
     Input:  wp,wc: The weighted antenna response fuctions to plus and cross
                    polarisations respectively
+            projection_type: standard, left or right
```
Comment (a-r-williamson) [inline bin/pycbc_multi_inspiral:208]: The kwarg here is `projection` not `projection_type`. Can you change the doc string?

File: bin/pycbc_multi_inspiral [hunk 8]
```diff
@@ -664,7 +711,9 @@ with ctx:
                 # entries that are single values need to be repeated once
                 # per event.
                 num_events = len(reweighted_snr)
+                # the output will only be possible if len(networkchi2) == num_events
 
+                logging.info("reweghtsnr, networkchi2 ={0}, {1}".format(num_events ,len(network_chisq_dict)))
```
Comment (a-r-williamson) [inline bin/pycbc_multi_inspiral:716]:
```suggestion
```
Not a particularly useful logging message for it to remain IMO

File: bin/pycbc_multi_inspiral [hunk 9]
```diff
@@ -590,22 +613,46 @@ with ctx:
...
+                                              coinc_triggers_r)
+                    rho_coinc = np.where(max_idx==0, rho_coinc_l, rho_coinc_r)
+                else:
+                    project = get_projection_matrix(wp, wc, projection=opt.projection)
+                    rho_coh, coinc_idx, coinc_triggers, rho_coinc = \
+                            coherent_snr(coinc_triggers, coinc_idx,
+                                         opt.coinc_threshold, project,
+                                         rho_coinc)
                 logging.info("%s points above coherent threshold"
                             % str(len(rho_coh)))
```
Comment (a-r-williamson) [inline bin/pycbc_multi_inspiral:657]:
```suggestion
                logging.info("%d points above coherent threshold", len(rho_coh))
```

File: bin/pycbc_multi_inspiral [hunk 10]
```diff
@@ -590,22 +613,46 @@ with ctx:
                 if len(coinc_idx) != 0:
                     logging.info("Max coincident SNR = %s"
                                 % str(max(rho_coinc)))
```
Comment (a-r-williamson) [inline bin/pycbc_multi_inspiral:615]:
```suggestion
                    logging.info("Max coincident SNR = %.2f", max(rho_coinc))
```

**Review by a-r-williamson — CHANGES_REQUESTED** [review]
Hi @JAMSADIQ, thanks for the recent commits. I've requested a few minor changes here, which hopefully you can just click accept on then I can approve the PR.

**Pull request merged**